File: src/agentbom/drift.py
"""Semantic drift analysis for AgentBOM snapshots."""

from __future__ import annotations

from dataclasses import dataclass
from enum import StrEnum
from typing import Iterable

from .snapshot import Snapshot, verify_snapshot


class DriftType(StrEnum):
    ADDED_ENTITY = "added_entity"
    REMOVED_ENTITY = "removed_entity"
    CHANGED_ENTITY = "changed_entity"
    ADDED_RELATIONSHIP = "added_relationship"
    REMOVED_RELATIONSHIP = "removed_relationship"


@dataclass(frozen=True, slots=True)
class DriftFinding:
    drift_type: DriftType
    severity: str
    title: str
    description: str
    key: str


SENSITIVE_KINDS = {"credential", "permission", "identity", "deployment", "database", "data_source"}
HIGH_IMPACT_RELATIONS = {"grants", "accesses", "authenticates_as", "assumes", "delegates", "writes", "calls"}


def compare_snapshots(previous: Snapshot, current: Snapshot) -> tuple[DriftFinding, ...]:
    if not verify_snapshot(previous):
        raise ValueError("Previous snapshot digest verification failed")
    if not verify_snapshot(current):
        raise ValueError("Current snapshot digest verification failed")

    findings: list[DriftFinding] = []
    previous_entities = {str(item["id"]): item for item in previous.entities}
    current_entities = {str(item["id"]): item for item in current.entities}
    previous_relationships = {_relationship_key(item): item for item in previous.relationships}
    current_relationships = {_relationship_key(item): item for item in current.relationships}

    for key in sorted(current_entities.keys() - previous_entities.keys()):
        entity = current_entities[key]
        kind = str(entity.get("kind", "unknown"))
        severity = "high" if kind in SENSITIVE_KINDS else "medium"
        findings.append(
            DriftFinding(
                DriftType.ADDED_ENTITY,
                severity,
                f"New {kind} appeared",
                f"Entity {entity.get('name', key)} was not present in the previous baseline.",
                key,
            )
        )

    for key in sorted(previous_entities.keys() - current_entities.keys()):
        entity = previous_entities[key]
        findings.append(
            DriftFinding(
                DriftType.REMOVED_ENTITY,
                "low",
                f"{entity.get('kind', 'entity').title()} removed",
                f"Entity {entity.get('name', key)} is no longer present.",
                key,
            )
        )

    for key in sorted(current_entities.keys() & previous_entities.keys()):
        before = previous_entities[key]
        after = current_entities[key]
        if before != after:
            kind = str(after.get("kind", "unknown"))
            severity = "high" if kind in SENSITIVE_KINDS else "medium"
            findings.append(
                DriftFinding(
                    DriftType.CHANGED_ENTITY,
                    severity,
                    f"{kind.title()} changed",
                    f"Entity {after.get('name', key)} changed since the previous baseline.",
                    key,
                )
            )

    for key in sorted(current_relationships.keys() - previous_relationships.keys()):
        relationship = current_relationships[key]
        relation = str(relationship.get("kind", "unknown"))
        severity = "high" if relation in HIGH_IMPACT_RELATIONS else "medium"
        findings.append(
            DriftFinding(
                DriftType.ADDED_RELATIONSHIP,
                severity,
                "New security relationship appeared",
                f"Relationship {relation} connects {relationship.get('source')} to {relationship.get('target')}.",
                key,
            )
        )

    for key in sorted(previous_relationships.keys() - current_relationships.keys()):
        relationship = previous_relationships[key]
        findings.append(
            DriftFinding(
                DriftType.REMOVED_RELATIONSHIP,
                "low",
                "Security relationship removed",
                f"Relationship {relationship.get('kind', 'unknown')} no longer connects {relationship.get('source')} to {relationship.get('target')}.",
                key,
            )
        )

    return tuple(findings)


def summarize_drift(findings: Iterable[DriftFinding]) -> dict[str, object]:
    findings = tuple(findings)
    counts = {severity: sum(1 for finding in findings if finding.severity == severity) for severity in ("critical", "high", "medium", "low")}
    return {"total": len(findings), "counts": counts}


def _relationship_key(item: dict[str, object]) -> str:
    return f"{item.get('source')}|{item.get('kind')}|{item.get('target')}"

File: rust/agentbom-engine/src/drift.rs
use agentbom_core::SecurityGraph;
use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
pub struct DriftFinding {
    pub drift_type: String,
    pub severity: String,
    pub title: String,
    pub description: String,
    pub key: String,
    #[serde(default)]
    pub evidence: Vec<String>,
}

pub fn analyze_drift(current: &SecurityGraph, baseline: &SecurityGraph) -> Vec<DriftFinding> {
    let diff = baseline.diff(current);
    let mut findings = Vec::new();

    for node in diff.added_nodes {
        let (severity, reason) = entity_severity(&node.kind, &node.name, &node.properties);
        findings.push(DriftFinding {
            drift_type: "added_entity".into(), severity: severity.into(),
            title: format!("New {} introduced", node.kind),
            description: format!("New entity '{}' was added to the security graph.", node.name),
            key: node.id,
            evidence: vec![format!("sensitivity_reason={reason}")],
        });
    }

    for edge in diff.added_edges {
        let target = current.nodes.get(&edge.target);
        let (severity, reason) = relationship_severity(&edge.kind, target);
        findings.push(DriftFinding {
            drift_type: "added_relationship".into(),
            severity: severity.into(),
            title: format!("New {} relationship", edge.kind),
            description: format!("{} -> {} now exists.", edge.source, edge.target),
            key: format!("{}:{}:{}", edge.source, edge.kind, edge.target),
            evidence: vec![format!("target_kind={}", target.map(|n| n.kind.as_str()).unwrap_or("unknown")), format!("sensitivity_reason={reason}")],
        });
    }

    for node in diff.removed_nodes {
        findings.push(DriftFinding {
            drift_type: "removed_entity".into(), severity: "low".into(),
            title: format!("{} removed", node.kind),
            description: format!("Entity '{}' is no longer present.", node.name),
            key: node.id,
            evidence: Vec::new(),
        });
    }

    for edge in diff.removed_edges {
        findings.push(DriftFinding {
            drift_type: "removed_relationship".into(), severity: "low".into(),
            title: format!("{} relationship removed", edge.kind),
            description: format!("{} -> {} is no longer present.", edge.source, edge.target),
            key: format!("{}:{}:{}", edge.source, edge.kind, edge.target),
            evidence: Vec::new(),
        });
    }

    findings.sort_by(|a, b| (&a.severity, &a.key).cmp(&(&b.severity, &b.key)));
    findings
}

fn entity_severity(kind: &str, name: &str, properties: &serde_json::Value) -> (&'static str, String) {
    if let Some(sensitivity) = properties.get("sensitivity").and_then(|v| v.as_str()) {
        let normalized = sensitivity.to_ascii_lowercase();
        if normalized == "critical" { return ("critical", "explicit sensitivity metadata".into()); }
        if normalized == "high" { return ("high", "explicit sensitivity metadata".into()); }
    }
    match kind {
        "credential" | "permission" | "identity" => return ("high", format!("sensitive entity kind: {kind}")),
        "database" | "deployment" | "data_source" => {
            if ["prod", "production", "secret", "credential", "payment"].iter().any(|token| name.to_ascii_lowercase().contains(token)) {
                return ("critical", "name heuristic matched sensitive token".into());
            }
            return ("high", format!("high-impact entity kind: {kind}"));
        }
        "tool" => return ("high", "tool capability can change authority".into()),
        _ => {}
    }
    ("medium", "default entity classification".into())
}

fn relationship_severity(kind: &str, target: Option<&agentbom_core::Node>) -> (&'static str, String) {
    let (target_severity, target_reason) = target.map(|n| entity_severity(&n.kind, &n.name, &n.properties)).unwrap_or(("medium", "unknown target".into()));
    if matches!(kind, "grants" | "accesses" | "assumes" | "delegates" | "writes" | "reads" | "calls") {
        match target_severity {
            "critical" => return ("critical", format!("high-impact relationship to critical target; {target_reason}")),
            "high" => return ("high", format!("high-impact relationship to sensitive target; {target_reason}")),
            _ => return ("high", "high-impact relationship".into()),
        }
    }
    ("medium", format!("relationship kind {kind}"))
}
